File: apps/api/app/schemas/market_timeline.py
from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


class MarketTimelineItem(BaseModel):
    timestamp: datetime
    type: str
    title: str
    description: str
    source: str
    url: str | None = None
    severity: str | None = None
    status: str | None = None
    data: dict[str, Any] = Field(default_factory=dict)


class MarketTimelineRead(BaseModel):
    market_id: int
    items: list[MarketTimelineItem]

File: apps/web/app/lib/marketTimeline.ts
"use client";

import { API_BASE_URL } from "./api";

export type MarketTimelineItem = {
  timestamp: string;
  type: string;
  title: string;
  description: string;
  source: string;
  url?: string | null;
  severity?: string | null;
  status?: string | null;
  data: Record<string, unknown>;
};

export type MarketTimelineResponse = {
  market_id: number;
  items: MarketTimelineItem[];
};

export async function fetchMarketTimeline(
  marketId: number | string,
): Promise<MarketTimelineResponse> {
  const response = await fetch(`${API_BASE_URL}/markets/${marketId}/timeline`, {
    cache: "no-store",
  });

  if (!response.ok) {
    throw new Error(`/markets/${marketId}/timeline responded ${response.status}`);
  }

  return response.json() as Promise<MarketTimelineResponse>;
}
